python/hail/expr/expressions/typed_expressions.py

```diff
@@ -1188,23 +1238,16 @@ def annotate(self, **named_exprs):
         :class:`.StructExpression`
             Struct with new or updated fields.
         """
-        names = []
-        types = []
-        for f, t in self.dtype.items():
-            names.append(f)
-            types.append(t)
-        kwargs_struct = hl.struct(**named_exprs)
+        new_types = {n: t for (n, t) in self.dtype.items()}
 
-        for f, t in kwargs_struct.dtype.items():
-            if not f in self._fields:
-                names.append(f)
-                types.append(t)
+        for f, e in named_exprs.items():
+            new_types[f] = e.dtype
```

Comment:
I believe this is correct, but want to verify that `InsertFields({a: int, b: str}, {b: float, a: call})` gives you `{a: call, b: float}`. Is that right?

Reply:
(I think we should change these semantics but we need ot match the scala definitions for now)

Reply:
yeah.